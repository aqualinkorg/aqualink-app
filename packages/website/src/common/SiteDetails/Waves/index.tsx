import React from "react";
import {
  createStyles,
  Theme,
  Card,
  CardContent,
  Typography,
  Grid,
  makeStyles,
} from "@material-ui/core";
import classNames from "classnames";
import { isNil } from "lodash";

import UpdateInfo from "../../UpdateInfo";
import type { LatestDataASSofarValue } from "../../../store/Sites/types";
import { formatNumber } from "../../../helpers/numberUtils";
import { toRelativeTime } from "../../../helpers/dates";
import waves from "../../../assets/waves.svg";
import arrow from "../../../assets/directioncircle.svg";
import wind from "../../../assets/wind.svg";
import { styles as incomingStyles } from "../styles";

const Waves = ({ data, hasSpotter }: WavesProps) => {
  const {
    significantWaveHeight,
    waveMeanDirection,
    waveMeanPeriod,
    windSpeed,
    windDirection,
  } = data;

  const waveHeight = significantWaveHeight;

  // Make sure to get the direction the wind is COMING FROM.
  // use `numberUtils.invertDirection` if needed.
  const windDirectionFrom = windDirection?.value;
  const waveDirectionFrom = waveMeanDirection?.value;
  const classes = useStyles({
    windDirection: windDirectionFrom,
    wavesDirection: waveDirectionFrom,
  });

  const windRelativeTime =
    windSpeed?.timestamp && toRelativeTime(windSpeed.timestamp);

  return (
    <Card className={classes.root}>
      <CardContent className={classes.contentWrapper}>
        <Grid
          className={classes.content}
          container
          justify="center"
          alignContent="space-between"
          item
          xs={12}
        >
          <Grid className={classes.paddingContainer} container item xs={12}>
            <Typography
              className={`${classes.cardTitle} ${classes.coloredText}`}
              variant="h6"
            >
              WIND
            </Typography>
            <img className={classes.titleImages} alt="wind" src={wind} />
          </Grid>
          <Grid className={classes.paddingContainer} container item xs={12}>
            <Grid item xs={6}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                SPEED
              </Typography>
              <Grid container alignItems="baseline">
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {/* Transform wind speed from m/s to km/h */}
                  {formatNumber(windSpeed?.value && windSpeed.value * 3.6, 1)}
                </Typography>
                {windSpeed?.value.toString() && (
                  <Typography
                    className={classes.contentUnits}
                    color="textSecondary"
                    variant="h6"
                  >
                    km/h
                  </Typography>
                )}
              </Grid>
            </Grid>
            <Grid item xs={6}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                DIRECTION
              </Typography>
              <Grid container alignItems="baseline">
                {!isNil(windDirectionFrom) && (
                  <img
                    className={classNames(
                      classes.arrow,
                      classes.windDirectionArrow
                    )}
                    alt="arrow"
                    src={arrow}
                  />
                )}
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {!isNil(windDirectionFrom)
                    ? `${formatNumber(windDirectionFrom)}\u00B0`
                    : "- -"}
                </Typography>
              </Grid>
            </Grid>
          </Grid>
          <Grid className={classes.paddingContainer} container item xs={12}>
            <Typography
              className={`${classes.cardTitle} ${classes.coloredText}`}
              variant="h6"
            >
              WAVES
            </Typography>
            <img className={classes.titleImages} alt="waves" src={waves} />
          </Grid>
          <Grid
            className={classes.paddingContainer}
            item
            xs={12}
            container
            justify="space-between"
          >
            <Grid item lg={4}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                HEIGHT
              </Typography>
              <Grid container alignItems="baseline">
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {formatNumber(waveHeight?.value, 1)}
                </Typography>
                {!isNil(waveHeight?.value) && (
                  <Typography
                    className={classes.contentUnits}
                    color="textSecondary"
                    variant="h6"
                  >
                    m
                  </Typography>
                )}
              </Grid>
            </Grid>
            <Grid item lg={3}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                PERIOD
              </Typography>
              <Grid container alignItems="baseline">
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {formatNumber(waveMeanPeriod?.value)}
                </Typography>
                {!isNil(waveMeanPeriod?.value) && (
                  <Typography
                    className={classes.contentUnits}
                    color="textSecondary"
                    variant="h6"
                  >
                    s
                  </Typography>
                )}
              </Grid>
            </Grid>
            <Grid item lg={5}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                DIRECTION
              </Typography>
              <Grid container alignItems="baseline">
                {!isNil(waveDirectionFrom) && (
                  <img
                    className={classNames(
                      classes.arrow,
                      classes.wavesDirectionArrow
                    )}
                    alt="arrow"
                    src={arrow}
                  />
                )}
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {!isNil(waveDirectionFrom)
                    ? `${formatNumber(waveDirectionFrom)}\u00B0`
                    : "- -"}
                </Typography>
              </Grid>
            </Grid>
          </Grid>
          <UpdateInfo
            relativeTime={windRelativeTime}
            timeText={hasSpotter ? "Last data received" : "Forecast valid for"}
            live={hasSpotter}
            frequency={hasSpotter ? "hourly" : "every 6 hours"}
            href="https://www.ncdc.noaa.gov/data-access/model-data/model-datasets/global-forcast-system-gfs"
            imageText={hasSpotter ? undefined : "NOAA GFS"}
          />
        </Grid>
      </CardContent>
    </Card>
  );
};

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    ...incomingStyles,
    root: {
      height: "100%",
      width: "100%",
      backgroundColor: "#eff0f0",
      display: "flex",
      flexDirection: "column",
    },
    coloredText: {
      color: theme.palette.primary.main,
    },
    titleImages: {
      height: 24,
      marginLeft: "0.5rem",
    },
    paddingContainer: {
      padding: "0.5rem 1rem",
    },
    contentWrapper: {
      height: "100%",
      flex: "1 1 auto",
      padding: 0,
    },
    content: {
      height: "100%",
    },
    arrow: {
      width: 20,
      height: 20,
      marginRight: "1rem",
      marginBottom: 10,
      [theme.breakpoints.between("md", 1350)]: {
        width: 15,
        height: 15,
      },
    },
    windDirectionArrow: ({ windDirection }: StyleProps) => ({
      transform: `rotate(${(windDirection || 0) + 180}deg)`,
    }),
    wavesDirectionArrow: ({ wavesDirection }: StyleProps) => ({
      transform: `rotate(${(wavesDirection || 0) + 180}deg)`,
    }),
  })
);

interface WavesProps {
  data: LatestDataASSofarValue;
  hasSpotter: boolean;
}

interface StyleProps {
  windDirection?: number;
  wavesDirection?: number;
}

export default Waves;
