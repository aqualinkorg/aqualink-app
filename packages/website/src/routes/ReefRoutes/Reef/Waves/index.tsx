import React from "react";
import {
  withStyles,
  WithStyles,
  createStyles,
  Theme,
  Card,
  CardContent,
  Typography,
  Grid,
} from "@material-ui/core";

import type { LiveData } from "../../../../store/Reefs/types";
import { formatNumber } from "../../../../helpers/numberUtils";
import waves from "../../../../assets/waves.svg";
import arrow from "../../../../assets/directioncircle.svg";
import wind from "../../../../assets/wind.svg";
import { styles as incomingStyles } from "../styles";

const Waves = ({ liveData, classes }: WavesProps) => {
  const {
    waveHeight,
    waveDirection,
    wavePeriod,
    windSpeed,
    windDirection,
  } = liveData;

  return (
    <Card className={classes.card}>
      <CardContent className={classes.contentWrapper}>
        <Grid
          className={classes.content}
          container
          justify="center"
          item
          xs={12}
        >
          <Grid container item xs={12}>
            <Typography className={classes.cardTitle} variant="h6">
              WIND
            </Typography>
            <img className={classes.titleImages} alt="wind" src={wind} />
          </Grid>
          <Grid container justify="flex-start" item xs={12}>
            <Grid item xs={6}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                SPEED
              </Typography>
              <Grid container alignItems="baseline">
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {formatNumber(windSpeed?.value, 1)}
                </Typography>
                {windSpeed?.value && (
                  <Typography
                    className={classes.contentUnits}
                    color="textSecondary"
                    variant="h6"
                  >
                    km/h
                  </Typography>
                )}
              </Grid>
            </Grid>
            <Grid item xs={6}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                DIRECTION
              </Typography>
              <Grid container alignItems="baseline">
                {windDirection?.value && (
                  <img
                    style={{
                      transform: `rotate(${windDirection?.value + 180}deg)`,
                    }}
                    className={classes.arrow}
                    alt="arrow"
                    src={arrow}
                  />
                )}
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {windDirection?.value
                    ? `${formatNumber(windDirection?.value)}\u00B0`
                    : "- -"}
                </Typography>
              </Grid>
            </Grid>
          </Grid>
          <Grid container item xs={12}>
            <Typography className={classes.cardTitle} variant="h6">
              WAVES
            </Typography>
            <img className={classes.titleImages} alt="waves" src={waves} />
          </Grid>
          <Grid item xs={12} container justify="space-between">
            <Grid item lg={4}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                HEIGHT
              </Typography>
              <Grid container alignItems="baseline">
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {formatNumber(waveHeight?.value, 1)}
                </Typography>
                {waveHeight?.value && (
                  <Typography
                    className={classes.contentUnits}
                    color="textSecondary"
                    variant="h6"
                  >
                    m
                  </Typography>
                )}
              </Grid>
            </Grid>
            <Grid item lg={3}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                PERIOD
              </Typography>
              <Grid container alignItems="baseline">
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {formatNumber(wavePeriod?.value)}
                </Typography>
                {wavePeriod?.value && (
                  <Typography
                    className={classes.contentUnits}
                    color="textSecondary"
                    variant="h6"
                  >
                    s
                  </Typography>
                )}
              </Grid>
            </Grid>
            <Grid item lg={5}>
              <Typography
                className={classes.contentTextTitles}
                color="textSecondary"
                variant="subtitle2"
              >
                DIRECTION
              </Typography>
              <Grid container alignItems="baseline">
                {waveDirection?.value && (
                  <img
                    style={{
                      transform: `rotate(${waveDirection?.value + 180}deg)`,
                    }}
                    className={classes.arrow}
                    alt="arrow"
                    src={arrow}
                  />
                )}
                <Typography
                  className={classes.contentTextValues}
                  color="textSecondary"
                  variant="h3"
                >
                  {formatNumber(waveDirection?.value)}
                  {waveDirection?.value ? "\u00B0" : ""}
                </Typography>
              </Grid>
            </Grid>
          </Grid>
        </Grid>
      </CardContent>
    </Card>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    ...incomingStyles,
    card: {
      ...incomingStyles.card,
      height: "100%",
      width: "100%",
      backgroundColor: "#eff0f0",
      display: "flex",
      flexDirection: "column",
    },
    cardTitle: {
      lineHeight: 1.5,
      color: theme.palette.primary.main,
    },
    titleImages: {
      height: 24,
      marginLeft: "0.5rem",
    },
    contentWrapper: {
      height: "100%",
      flex: "1 1 auto",
      padding: 0,
    },
    content: {
      padding: "1rem",
      height: "100%",
    },
    arrow: {
      width: 20,
      height: 20,
      marginRight: "1rem",
      marginBottom: 10,
      [theme.breakpoints.between("md", 1350)]: {
        width: 15,
        height: 15,
      },
    },
  });

interface WavesIncomingProps {
  liveData: LiveData;
}

type WavesProps = WithStyles<typeof styles> & WavesIncomingProps;

export default withStyles(styles)(Waves);
