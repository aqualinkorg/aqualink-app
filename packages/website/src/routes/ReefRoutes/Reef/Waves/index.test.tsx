import React from "react";
import { render } from "@testing-library/react";

import Waves from ".";

test("renders as expected", () => {
  const liveData = {
    reef: { id: 1 },
    date: "2020-07-01T14:25:18.008Z",
    bottomTemperature: {
      value: 25,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
    satelliteTemperature: {
      value: 26,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
    degreeHeatingDays: {
      value: 29,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
    waveHeight: {
      value: 3,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
    waveDirection: {
      value: 136,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
    wavePeriod: {
      value: 15,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
    windSpeed: {
      value: 5,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
    windDirection: {
      value: 96,
      timestamp: "2020-07-01T14:25:18.008Z",
    },
  };
  const { container } = render(<Waves liveData={liveData} />);
  expect(container).toMatchSnapshot();
});
