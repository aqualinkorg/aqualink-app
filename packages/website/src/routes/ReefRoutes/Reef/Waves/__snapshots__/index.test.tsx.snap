// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <mock-card
    classname="Waves-card-1"
  >
    <mock-cardcontent
      classname="Waves-contentWrapper-10"
    >
      <div
        class="MuiGrid-root Waves-content-11 MuiGrid-container MuiGrid-item MuiGrid-align-content-xs-space-between MuiGrid-justify-xs-center MuiGrid-grid-xs-12"
      >
        <div
          class="MuiGrid-root Waves-paddingContainer-9 MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12"
        >
          <mock-typography
            classname="Waves-cardTitle-2"
            variant="h6"
          >
            WIND
          </mock-typography>
          <img
            alt="wind"
            class="Waves-titleImages-8"
            src="wind.svg"
          />
        </div>
        <div
          class="MuiGrid-root Waves-paddingContainer-9 MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <mock-typography
              classname="Waves-contentTextTitles-4"
              color="textSecondary"
              variant="subtitle2"
            >
              SPEED
            </mock-typography>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-baseline"
            >
              <mock-typography
                classname="Waves-contentTextValues-5"
                color="textSecondary"
                variant="h3"
              >
                5.0
              </mock-typography>
              <mock-typography
                classname="Waves-contentUnits-6"
                color="textSecondary"
                variant="h6"
              >
                km/h
              </mock-typography>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <mock-typography
              classname="Waves-contentTextTitles-4"
              color="textSecondary"
              variant="subtitle2"
            >
              DIRECTION
            </mock-typography>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-baseline"
            >
              <img
                alt="arrow"
                class="Waves-arrow-12"
                src="directioncircle.svg"
                style="transform: rotate(276deg);"
              />
              <mock-typography
                classname="Waves-contentTextValues-5"
                color="textSecondary"
                variant="h3"
              >
                96°
              </mock-typography>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root Waves-paddingContainer-9 MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12"
        >
          <mock-typography
            classname="Waves-cardTitle-2"
            variant="h6"
          >
            WAVES
          </mock-typography>
          <img
            alt="waves"
            class="Waves-titleImages-8"
            src="waves.svg"
          />
        </div>
        <div
          class="MuiGrid-root Waves-paddingContainer-9 MuiGrid-container MuiGrid-item MuiGrid-justify-xs-space-between MuiGrid-grid-xs-12"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-4"
          >
            <mock-typography
              classname="Waves-contentTextTitles-4"
              color="textSecondary"
              variant="subtitle2"
            >
              HEIGHT
            </mock-typography>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-baseline"
            >
              <mock-typography
                classname="Waves-contentTextValues-5"
                color="textSecondary"
                variant="h3"
              >
                3.0
              </mock-typography>
              <mock-typography
                classname="Waves-contentUnits-6"
                color="textSecondary"
                variant="h6"
              >
                m
              </mock-typography>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-3"
          >
            <mock-typography
              classname="Waves-contentTextTitles-4"
              color="textSecondary"
              variant="subtitle2"
            >
              PERIOD
            </mock-typography>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-baseline"
            >
              <mock-typography
                classname="Waves-contentTextValues-5"
                color="textSecondary"
                variant="h3"
              >
                15
              </mock-typography>
              <mock-typography
                classname="Waves-contentUnits-6"
                color="textSecondary"
                variant="h6"
              >
                s
              </mock-typography>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-5"
          >
            <mock-typography
              classname="Waves-contentTextTitles-4"
              color="textSecondary"
              variant="subtitle2"
            >
              DIRECTION
            </mock-typography>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-baseline"
            >
              <img
                alt="arrow"
                class="Waves-arrow-12"
                src="directioncircle.svg"
                style="transform: rotate(316deg);"
              />
              <mock-typography
                classname="Waves-contentTextValues-5"
                color="textSecondary"
                variant="h3"
              >
                136
                °
              </mock-typography>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root UpdateInfo-updateInfo-13 false MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
            >
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root UpdateInfo-updateIcon-15 MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M21 10.12h-6.78l2.74-2.82c-2.73-2.7-7.15-2.8-9.88-.1-2.73 2.71-2.73 7.08 0 9.79s7.15 2.71 9.88 0C18.32 15.65 19 14.08 19 12.1h2c0 1.98-.88 4.55-2.64 6.29-3.51 3.48-9.21 3.48-12.72 0-3.5-3.47-3.53-9.11-.02-12.58s9.14-3.47 12.65 0L21 3v7.12zM12.5 8v4.25l3.5 2.08-.72 1.21L11 13V8h1.5z"
                  />
                </svg>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <mock-box
                  display="flex"
                  flexdirection="column"
                >
                  <mock-typography
                    classname="UpdateInfo-updateInfoText-16"
                    variant="caption"
                  >
                    Last data received
                     
                    5 min. ago
                  </mock-typography>
                  <mock-typography
                    classname="UpdateInfo-updateInfoText-16"
                    variant="caption"
                  >
                    Updated 
                    hourly
                  </mock-typography>
                </mock-box>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root UpdateInfo-nooaChip-17 MuiGrid-item"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
            >
              <div
                class="UpdateInfo-circle-21"
              />
              <mock-typography
                classname="UpdateInfo-nooaChipText-19"
              >
                LIVE
              </mock-typography>
            </div>
          </div>
        </div>
      </div>
    </mock-cardcontent>
  </mock-card>
</div>
`;
